src/main/java/org/truffleruby/extra/ConcurrentHashMapNodes.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright (c) 2015, 2020 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.extra;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.object.Shape;
+import org.truffleruby.builtins.CoreMethod;
+import org.truffleruby.builtins.CoreMethodArrayArgumentsNode;
+import org.truffleruby.builtins.CoreModule;
+import org.truffleruby.builtins.YieldingCoreMethodNode;
+import org.truffleruby.core.basicobject.RubyBasicObject;
+import org.truffleruby.core.hash.RubyHash;
+import org.truffleruby.core.klass.RubyClass;
+import org.truffleruby.core.proc.RubyProc;
+import org.truffleruby.language.Nil;
+import org.truffleruby.language.NotProvided;
+import org.truffleruby.language.Visibility;
+import org.truffleruby.language.objects.AllocationTracing;
+
+import java.util.Map;
+
+@CoreModule(value = "TruffleRuby::ConcurrentHashMap", isClass = true)
```

Comment:
Naming-wise, should it be `TruffleRuby::ConcurrentHashMap` or `TruffleRuby::ConcurrentMap`?
It does not preserve insertion ordering (just like Concurrent::Map, unlike Concurrent::Hash), so maybe avoiding `Hash` in the name might be clearer, and it'd be more clearly related to `Concurrent::Map`.

OTOH I can see that `TruffleRuby::ConcurrentHashMap` is explicit that it uses a Java ConcurrentHashMap behind the scenes, but I feel that doesn't make so much sense to expose in the Ruby class name.

Reply:
Could call it `ConcurrentDict` to make it super clear it's not a `Hash`?

Reply:
That sounds like Python :laughing:  I think `TruffleRuby::ConcurrentMap` is what captures it best here, let's use ConcurrentRuby's terminology which is well established.

Reply:
Renamed `TruffleRuby::ConcurrentHashMap` to `TruffleRuby::ConcurrentMap` in [`f186aad` (#2339)](https://github.com/oracle/truffleruby/pull/2339/commits/f186aad6a1397ab6483798d5c02f112a1b57cd9e)